tools/cbmc/proofs/Makefile.template

```diff
@@ -73,14 +63,27 @@ unpatch:
 #
 # Rules for building the CBMC harness
 
-queue_datastructure.h: $(H_OBJS_EXCEPT_HARNESS)
-	python3 @TYPE_HEADER_SCRIPT@ --binary $(FREERTOS)/freertos_kernel/queue.goto --c-file $(FREERTOS)/freertos_kernel/queue.c
+C_SOURCES = $(patsubst %.goto,%.c,$(H_OBJS_EXCEPT_HARNESS))
 
-$(ENTRY)_harness.goto: $(ENTRY)_harness.c $(H_GENERATE_HEADER)
+# Build each goto-binary out-of-source (i.e. in a 'gotos' directory
+# underneath each proof directory, to make it safe to build all proofs
+# in parallel
+OOS_OBJS = $(patsubst %.c,gotos/%.goto,$(C_SOURCES))
+
+CWD=$(abspath .)
+
+gotos/%.goto: %.c
+	mkdir -p $(dir $@)
 	$(GOTO_CC) @COMPILE_ONLY@ @RULE_OUTPUT@ $(INC) $(CFLAGS) @RULE_INPUT@
 
-$(ENTRY)1.goto: $(OBJS)
-	$(GOTO_CC) @COMPILE_LINK@ @RULE_OUTPUT@ --function harness $(OBJS)
+queue_datastructure.h: gotos/$(FREERTOS)/freertos_kernel/queue.goto
+	python3 @TYPE_HEADER_SCRIPT@ --binary $(CWD)/gotos$(FREERTOS)/freertos_kernel/queue.goto --c-file $(FREERTOS)/freertos_kernel/queue.c
```

Comment:
* There should be a `/` in `$(CWD)/gotos$(FREERTOS)` between `gotos` and `$(FREERTOS)`
* Currently, the prerequisite is `gotos/$(FREERTOS)/...` but the argument is `$(CWD)/gotos/...`. I would prefer if the argument was just `$@` so that we don't write the same thing twice and have it go out-of-date. I believe there's no need for `$(CWD)` because this command will be run in the proof directory, but IMO you should still have it there just so that an absolute path shows up in the log.

Reply:
I have addressed all of these issues with `--binary @RULE_INPUT@`.  This required modifying the type header script to expand this relative path to an absolute patch before running the goto-instrument command that uses this path.  The script changes directory to a temporary directory before running goto-instrument, so an absolute path is required, probably explaining  the reason for the $(CWD) prefix used in the prior version of this Makefile.template.